BUIP055 partial implementation

gandrewstone opened the pull request:
Implements:
fork trigger
opt-in replay protection for 1MB chain txs

gandrewstone committed:
BUIP055 implementation including fork control and invalid OP_RETURN transaction, and relevant tests.

ftrader committed:
Add doc/buip-hf.md guidance document

ftrader committed:
Fix obsoleted link to tech spec

ptschip committed:
Add 4 block attack scenarios to PV python testing

Also add the needed option -pvtest which slows down the checking
of inputs by putting a sleep time of 1 second for each input to
check.  This allows up to simulate very long to validate blocks
used in the attack scenarios.

ptschip committed:
Check for reorgs in progress when new blocks are beginning validation.

Remove unnecessary code

gandrewstone committed:
fix rare test error

ptschip committed:
Move block relay to ActivateBestChainStep

In PV we can connect blocks beyone pindexMostWork if other
blocks arrive just after we start validation.  Therefore we
need to relay block inventory based on the new pindexMostWork
after all the blocks have been connected.  If we instead
return to ActivateBestChain we will relay inventory based on the
old pindexMostWork and may end up not relaying a block.

ptschip committed:
Indent section

No changes to logic, just indentation and comment changes.

ptschip committed:
Add extended test for multiple block validation

Mine  blocks from different nodes at the same time repeatedly
for 100 iterations.

ptschip commented:
needs a rebase

ftrader commented on doc/buip-hf.md:
```diff
@@ -0,0 +1,103 @@
+BUIP-HF Development Notes
```
Most of this document was applicable to working on my personal repo.

I think this document does not need to remain, or can be trimmed down do bare essentials as the work continues on BU's `dev` branch.

sickpig commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -150,28 +151,34 @@ def hex_str_to_bytes(hex_str):
     while True:
         pool = set(rpc_connections[0].getrawmempool())
         num_match = 1
+        poolLen = [len(pool)]
```
nit pick: why using camelcase here?

sickpig commented on src/main.h:
```diff
@@ -113,7 +113,7 @@ static const int64_t BLOCK_DOWNLOAD_TIMEOUT_BASE = 1000000;
 static const int64_t BLOCK_DOWNLOAD_TIMEOUT_PER_PEER = 500000;
 /** Timeout in secs for the initial sync. If we don't receive the first batch of headers */
-static const uint32_t INITIAL_HEADERS_TIMEOUT = 30;
+static const uint32_t INITIAL_HEADERS_TIMEOUT = 120;
```
why is this change needed?

gandrewstone replied:
buip055 test was failing because of timeout on my test setup, when I used a rotating disk.  But working fine using a ramdisk.  I talked to @ptschip and we agreed that its ok to increase this value.

sickpig replied:
mmm ok, was just worried by the fact that the change does not seem related to the PR

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+}
+
+bool ValidateBUIP055Tx(const CTransaction& tx)
+{
```
I cant's see anywhere the code that implement the sunset rule that consider this kind of txn invalid only if block height <= 530000, see REQ-6.1.

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+    for (const CTransaction &tx : block.vtx)
+    {
+        if (!ValidateBUIP055Tx(tx))
+            return state.DoS(100,
```
Why using 100, that means that the peer will be dos banned as soon as we see a txn that activate the replay protection, is this really what we want?

gandrewstone replied:
in this case, a peer has given us a block with a bad tx in it.  The tx has replay protection, but it should not.  So it has mis-coded the BUIP055 rules.  I suppose we could not ban it, just ignore the block...

sickpig replied:
no strong opinion, I wonder what other think? paging @ftrader @ptschip @deadalnix @dagurval @dgenr8

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+#include <queue>
+
+std::vector<unsigned char> invalidOpReturn = {112, 101, 101, 114, 50, 112, 101, 101, 114, 32, 99, 97, 115, 104, 32, 114, 101, 113, 117, 105, 114, 101, 115, 32, 108, 97, 114, 103, 101, 114, 32, 98, 108, 111, 99, 107, 115};
+
```
why not just using an `static const std::string` ("peer2peer cash requires larger blocks") rather than a vector of unsigned char and convert the it to a vector just before the comparison in `IsTxOpReturnInvalid`? 
IMHO having `invalidOpReturn` stated as a string would make the code clearer.

dgenr8 commented on src/main.cpp:
```diff
@@ -1154,6 +1155,16 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
 
+    // BUIP055: reject transactions that won't work on the fork.
+    // This code uses the system time to determine when to start rejecting which is inaccurate relative to the
+    // actual activation time (defined by times in the blocks).
```
Could chainActive.Tip()->GetMedianTimePast() be used?

dgenr8 commented on src/main.cpp:
```diff
@@ -4042,6 +4054,26 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
+    // BUIP055 check soft-fork items, such as tx targeted to the 1MB chain
+    if (pindexPrev && pindexPrev->IsforkActiveOnNextBlock(miningForkTime.value))
+    {
+        return ValidateBUIP055Block(block, state);
```
Could this be done inside the existing tx loop in ContextualCheckBlock, rather than a new tx loop?

ptschip committed:
Add extended reorg tests for PV

There are many different reorg possibilities that are tested
in this addition to the python tests.

gandrewstone committed:
Merge branch 'dev' into buip-hf

ptschip committed:
Fix edge case with multi reorgs

Handle a reorg that is not yet complete but another reorg
is initiated.

ptschip committed:
Run clang-format on main.cpp

gandrewstone committed:
fix excessive.py extended tests to work faster and with the xthin excessive size changes

ptschip committed:
return if pindexBestHeader is NULL

FindMostWorkChain() can return an NULL.

Also updated fix for compiler warning. Casting pindex->nHeight to
uint32_t.  TODO: why is pindex->nHeight an int, it really should
be uint anway.

Add Parenthesis for code clarity.

gandrewstone committed:
Merge pull request #696 from ptschip/dev_pv

PV: reorg bug fix (issue #566)

gandrewstone committed:
Merge pull request #704 from gandrewstone/qa_largeTxns

fix excessive.py extended tests to work faster and with the xthin exc…

sickpig commented:
@gandrewstone this PR need a rebase

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017 The Bitcoin Unlimited Developers
```
wonder if we should call it `uahffork.cpp` or `bitcashfork.cpp`

sickpig commented on src/buip055fork.h:
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2017 The Bitcoin Unlimited Developers
```
same here we should consider to rename the file

sickpig committed:
Add sunset clause to REQ-6-1 implementation.

As per UAHF spec (*) section REQ-6-1 once the fork has activated,
transactions containing an OP_RETURN output with a specific magic data
value of

`Bitcoin: A Peer-to-Peer Electronic Cash System`

(46 characters, including the single spaces separating the words, and
without any terminating null character) shall be considered invalid
until block 530,000 inclusive.

This commit just add the sunset clause, i.e. block height lower than
530,001.

(*) https://github.com/Bitcoin-UAHF/spec/blob/master/uahf-technical-spec.md#req-6-1-disallow-special-op_return-marked-transactions-with-sunset-clause

sickpig committed:
Update the REQ-6-1 anti reply magic value

While at it use a string to define it. That way
is more clear and more to the point we do the
same also in the python tests.

gandrewstone committed:
Merge pull request #2 from sickpig/add-req-6-1-sunset-clause

Add sunset clause to REQ-6-1 implementation.

gandrewstone committed:
Merge pull request #3 from sickpig/update-anti-reply-magic-value

Update anti reply magic value

gandrewstone committed:
merge
